### PR TITLE
[missiontimer] fix for debug error on non-ready clients

### PR DESCRIPTION
**Pull request by Einheit-101**
New PR in this Branch

### CODE REVIEW

**Comment by Einheit-101** [comment]
Now i know how this works. I added all 4 script files.

**Comment by qaisjp** [comment]
Are you using the GitHub interface to make these pull requests? I suggest you learn how to use the Git command line to make your branches *maintainable*.

After your PR has been reviewed your branch should be squash merged, to prevent this from leaking into the main repository.

**Comment by Einheit-101** [comment]
I used the default web browser interface here.

File: [gameplay]/missiontimer/missiontimer_server.lua [hunk 1]
```diff
@@ -123,7 +128,18 @@ addEvent("onClientMissionTimerDownloaded",true)
 addEventHandler ( "onClientMissionTimerDownloaded", root, 
 	function()
 		for timer,data in pairs(missionTimers) do
-			triggerClientEvent ( source, "setupNewMissionTimer", timer, getMissionTimerTime(timer), data.countdown, data.timerFormat, data.x, data.y, data.bg, data.font, data.scale, data.r, data.g, data.b )
+			table.insert( readyPlayerList, client )
```
Comment (qaisjp) [inline [gameplay]/missiontimer/missiontimer_server.lua]: doesn't this add the player to the list for as many mission timers there are?

say you have two missionTimers. wouldn't be there be two clients in the `readyPlayerList`?

**Comment by Einheit-101** [comment]
Yup, I never have more than one MTs so I didn't think about that issue.... :D

File: [gameplay]/missiontimer/missiontimer_client.lua [hunk 2]
```diff
@@ -1,16 +1,14 @@
-﻿rootElement = getRootElement()
-thisResource = getThisResource()
 missionTimers = {}
 bool = { [false]=true, [true]=true }
 addEvent ( "onClientMissionTimerElapsed", true )
 
 addEventHandler("onClientResourceStart",resourceRoot,
 	function()
-		triggerServerEvent ( "onClientMissionTimerDownloaded", getLocalPlayer() )
+		triggerServerEvent ( "onClientMissionTimerDownloaded", root )
```
Comment (qaisjp) [inline [gameplay]/missiontimer/missiontimer_client.lua]: why was this changed from `localPlayer` to `root`?

Even though `source` is not used anymore (`client` is) the old code is more optimal (even though it's a tiny optimisation)

File: [gameplay]/missiontimer/missiontimer_server.lua [hunk 3]
```diff
@@ -1,18 +1,23 @@
...
+addEventHandler("onPlayerQuit",root,
+	function()
+		table.removevalue( readyPlayerList, source )
+	end
+)
+
 addEventHandler ("onResourceStop",root,
 	function()
-		for i,timer in ipairs(getElementsByType("missiontimer",source)) do
+		for i,timer in pairs(getElementsByType("missiontimer",source)) do
```
Comment (qaisjp) [inline [gameplay]/missiontimer/missiontimer_server.lua]: why has the iterator been changed?

File: [gameplay]/missiontimer/missiontimer_client.lua (hunk 2 above)
Comment (nokizorque) [inline [gameplay]/missiontimer/missiontimer_client.lua]: Wouldn't the correct usage here be `resourceRoot`?

File: [gameplay]/missiontimer/missiontimer_server.lua [hunk 4]
```diff
@@ -1,18 +1,23 @@
...
 			destroyElement(timer)
 		end
 	end
 )
 
 function createMissionTimer ( duration, countdown, timerFormat, x, y, bg, font, scale, r, g, b )
-	sourceResource = sourceResource or thisResource
+	sourceResource = sourceResource or resourceRoot
```
Comment (qaisjp) [inline [gameplay]/missiontimer/missiontimer_server.lua]: `thisResource` was `getThisResource()`, not `getResourceRoot()`

Should this be `sourceResource or resource`?

File: [gameplay]/missiontimer/missiontimer_server.lua [hunk 5]
```diff
@@ -55,7 +60,7 @@ function getMissionTimerTime ( timer )
 end
 
 function setMissionTimerFrozen ( timer, frozen )	
-	if not bool[frozen] then return false end
+	if not type(frozen) == "boolean" then return false end
```
Comment (qaisjp) [inline [gameplay]/missiontimer/missiontimer_server.lua]: use `type(frozen) ~= "boolean"`

File: [gameplay]/missiontimer/missiontimer_client.lua (hunk 2 above)
Comment (qaisjp) [inline [gameplay]/missiontimer/missiontimer_client.lua]: I believe resourceRoot would still propagate to objects created by the resource

Comment (ccw808) [inline [gameplay]/missiontimer/missiontimer_client.lua]: I think resourceRoot should be used here and also in the event handler on the server.

Comment (qaisjp) [inline [gameplay]/missiontimer/missiontimer_client.lua]: For the triggerServerEvent, or the onClientResourceStart?

Comment (ccw808) [inline [gameplay]/missiontimer/missiontimer_client.lua]: For triggerServerEvent onClientMissionTimerDownloaded (and addEventHandler onClientMissionTimerDownloaded)

Comment (ccw808) [inline [gameplay]/missiontimer/missiontimer_client.lua]: Reason is, using resourceRoot in both trigger and event handler, helps eliminate event name clashes.

Comment (Einheit-101) [inline [gameplay]/missiontimer/missiontimer_client.lua]: localPlayer in triggerServerEvent doesnt work for some reason so i replaced it with root after some testing (i wondered why just nothing happened). Didnt try resourceRoot tho. Maybe it works.

File: [gameplay]/missiontimer/missiontimer_server.lua (hunk 3 above)
Comment (Einheit-101) [inline [gameplay]/missiontimer/missiontimer_server.lua]: i thought pairs may be a slight bit faster and the order doesnt matter here.

File: [gameplay]/missiontimer/missiontimer_server.lua (hunk 4 above)
Comment (Einheit-101) [inline [gameplay]/missiontimer/missiontimer_server.lua]: If there is no other sourceResource, should sourceResource not automatically point to the missiontimer resource? That means, is this line even necessary?
sourceResource should never be nil

File: [gameplay]/missiontimer/missiontimer_server.lua (hunk 5 above)
Comment (Einheit-101) [inline [gameplay]/missiontimer/missiontimer_server.lua]: Yup better

**Comment by Einheit-101** [comment]
I am currently using the latest version with all recent changes i just made and it works.
triggerServerEvent ( "onClientMissionTimerDownloaded", localPlayer ) does nothing. resourceRoot or root is needed.

File: [gameplay]/missiontimer/missiontimer_server.lua (hunk 4 above)
Comment (qaisjp) [inline [gameplay]/missiontimer/missiontimer_server.lua]: Yeah, but resourceRoot refers to a `resource root element`, not an actual resource object

**Comment by qaisjp** [comment]
Use resourceRoot as ccw recommended

**Comment by Einheit-101** [comment]
Yes, done this.

**Pull request merged**